data/bookkeeping/txn_merkle.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package bookkeeping
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/crypto/merklearray"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// TxnMerkleTree returns a cryptographic commitment to the transactions in the
+// block, along with their ApplyData, as a Merkle tree.  This allows the
+// caller to either extract the root hash (for inclusion in the block
+// header), or to generate proofs of membership for transactions that are
+// in this block.
+func (block Block) TxnMerkleTree() (*merklearray.Tree, error) {
+	return merklearray.Build(&txnMerkleArray{block: block})
+}
+
+// txnMerkleArray is a representation of the transactions in this block,
+// along with their ApplyData, as an array for the merklearray package.
+type txnMerkleArray struct {
+	block Block
+}
+
+// Length implements the merklearray.Array interface.
+func (tma *txnMerkleArray) Length() uint64 {
+	return uint64(len(tma.block.Payset))
+}
+
+// Get implements the merklearray.Array interface.
+func (tma *txnMerkleArray) GetHash(pos uint64) (crypto.Digest, error) {
+	if pos >= uint64(len(tma.block.Payset)) {
+		return crypto.Digest{}, fmt.Errorf("txnMerkleArray.Get(%d) out of bounds %d", pos, len(tma.block.Payset))
```

Comment:
Super nit but consider changing the error to be more specific about the bounds. Maybe something along the lines of `txnMerkleArray.Get(%d) failed because you asked for transaction index %d but there are only %d transactions in the block` or `txnMerkleArray.Get(%d) out of bounds. Payset length - %d`

Reply:
I'm not quite sure what you're suggesting.  I believe the error message currently includes both the index requested (the first `%d`) as well as the expected bound (the second `%d`).  Are you thinking the wording should be more verbose?

Reply:
Yes, I think it can be tricky for someone to guess what is the second number if it's out of context. But again, very nit.